[online_image] Last-Modified-Date and ETag response caching

candrews opened the pull request:
# What does this implement/fix?

Use the Last-Modified-Date and ETag response headers (if present) when making subsequent requests to implement conditional response caching. If the data on the server is not modified, then a 304 Not Modified response results and the online_image component will reuse the previously received data instead of downloading and processing it again.

See: https://developer.mozilla.org/en-US/docs/Web/HTTP/Caching

The `on_download_finished` callback now has a boolean argument provided that is `true` when a new image is downloaded (cache miss); it's `false` when there is a cache hit.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- https://github.com/esphome/esphome-docs/pull/4906

## Test Environment

- [X] ESP32
- [X] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

candrews committed:
[online_image] Last-Modified-Date and ETag response caching

Use the Last-Modified-Date and ETag response headers (if present) when making subsequent requests to implement conditional response caching. If the data on the server is not modified, then a 304 Not Modified response results and the online_image component will reuse the previously received data instead of downloading and processing it again.

See: https://developer.mozilla.org/en-US/docs/Web/HTTP/Caching

jesserockz commented on esphome/components/online_image/online_image.h:
```diff
@@ -173,6 +173,14 @@ class OnlineImage : public PollingComponent,
    * decoded images).
    */
   int buffer_height_;
+  /**
+   * The value of the ETag HTTP header provided in the last response.
+   */
+  std::string etag_ = "";
+  /**
+   * The value of the Last-Modified HTTP header provided in the last response.
+   */
+  std::string last_modified_ = "";
```
While it wont make a huge difference, these should to be reset if the url is changed `set_url(...)` to allow the next request to be smaller and not send irrelevant headers

candrews replied:
Great point! I've done that: 5bf871dfd82bedc154be2416b90413f520993d36

jesserockz commented on esphome/components/http_request/http_request.cpp:
Added for debugging during development?

candrews replied:
I shouldn't have pushed that commit - I removed it.

candrews committed:
[online_image] reset last etag and last_modified when url changes

candrews committed:
[online_image] end_connect after reading response

candrews commented:
In my testing, this PR works great. I'm eager to get it merged, please let know what I can do. Thank you in advance for your review! :)

clydebarrow commented on esphome/components/online_image/online_image.cpp:
```diff
@@ -3,6 +3,10 @@
+static const std::string ETAG_HEADER_NAME = "etag";
+static const std::string IF_NONE_MATCH_HEADER_NAME = "if-none-match";
+static const std::string LAST_MODIFIED_HEADER_NAME = "last-modified";
+static const std::string IF_MODIFIED_SINCE_HEADER_NAME = "if-modified-since";
```
Declaring this way results in a copy of the string in flash, which is then used at startup to construct a `std::string` object in RAM, which means a copy of the data plus the class overhead permanently occupying RAM.

These should instead just be `const char * const` - when passed to a function requiring a `std::string` it will still be copied, but this will be a transient allocation instead of a permanent chunk of RAM being used up.

candrews replied:
I didn't know that, thank you explaining. I've made that change.

candrews committed:
[online_image] use char* instead of std::string for constants

std::string results in a copy of the string in flash, which is then used at startup to construct a std::string object in RAM, which means a copy of the data plus the class overhead permanently occupying RAM.

Instead use const char* const - when passed to a function requiring a std::string it will still be copied, but this will be a transient allocation instead of a permanent chunk of RAM being used up.
